Interpreter support for CallFunction/CallMethod

zdevito opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #21515 Add WeakIValue, use in tracer.
* #21514 clean up the TracingState API
* #21513 Collapse tracing_state.h into tracer.h
* **#21325 Interpreter support for CallFunction/CallMethod**
* #21273 Expose ExecutionPlan in prep for function calls
* #21272 Add flag to temporarily enable first class modules
* #21240 Reduce number of stack manipulation instructions in interpreter.
* #21185 Prepare interpreter for function calling

Summary: Adds CALL instruction, and function_table to interpreter state.
Interpreter state now has a list of Frame objects to keep track of
active call frames.

Nodes:
* Re-did how exception wrapping works so that call stacks can be printed
  by the interpreter.
* Change the order of error reporting so that it is more clear what
  part of a printout is attributable to the interpreter or shape prop
  adding additional context to an error.
* The first-class functions are only enabled under a guard because
  serialization does not work and more tests need to be added for
  the funnctionality.

Differential Revision: [D15618275](https://our.internmc.facebook.com/intern/diff/D15618275)

zdevito committed:
Interpreter support for CallFunction/CallMethod

Summary: Adds CALL instruction, and function_table to interpreter state.
Interpreter state now has a list of Frame objects to keep track of
active call frames.

Nodes:
* Re-did how exception wrapping works so that call stacks can be printed
  by the interpreter.
* Change the order of error reporting so that it is more clear what
  part of a printout is attributable to the interpreter or shape prop
  adding additional context to an error.
* The first-class functions are only enabled under a guard because
  serialization does not work and more tests need to be added for
  the funnctionality.

zdevito commented on torch/csrc/jit/interpreter.cpp:
```diff
@@ -611,173 +635,239 @@ struct InterpreterStateImpl : c10::intrusive_ptr_target {
+    ActiveFrame af(frames.back());
     try {
-      return runInterpreterLoop(stack);
+      while (true) {
```
This is just moving the interpreter loop inline here because the exception handling stuff needs to see `af.pc`. The ops themselves, except for CALL have not meaningfully changed.

zdevito committed:
Update on "Interpreter support for CallFunction/CallMethod"

Interpreter support for CallFunction/CallMethod

Summary: Adds CALL instruction, and function_table to interpreter state.
Interpreter state now has a list of Frame objects to keep track of
active call frames.

Nodes:
* Re-did how exception wrapping works so that call stacks can be printed
  by the interpreter.
* Change the order of error reporting so that it is more clear what
  part of a printout is attributable to the interpreter or shape prop
  adding additional context to an error.
* The first-class functions are only enabled under a guard because
  serialization does not work and more tests need to be added for
  the funnctionality.

gh-metadata: pytorch pytorch 21325 gh/zdevito/48/head

suo commented on torch/csrc/jit/interpreter.cpp:
```diff
@@ -661,173 +685,239 @@ struct InterpreterStateImpl : c10::intrusive_ptr_target {
+            const Code& code =
+                af.functions[inst.X]->get_executor().getPlanFor(stack).code;
+            frames.back().pc = af.pc + 1;
+            enterFrame(code);
```
I think `enterFrame` and `leaveFrame` should set the active frame. That would require `af` to be a member of the interpreter state though.

suo commented on torch/csrc/jit/exception_message.h:
```diff
@@ -0,0 +1,31 @@
+namespace torch {
+namespace jit {
+
+struct ExceptionMessage {
```
Can you add a comment explaining what purpose this serves? Why cut out the c10 backtarce is not obvious I think.

Commit:
Update on "Interpreter support for CallFunction/CallMethod"

Interpreter support for CallFunction/CallMethod

Summary: Adds CALL instruction, and function_table to interpreter state.
Interpreter state now has a list of Frame objects to keep track of
active call frames.

Nodes:
* Re-did how exception wrapping works so that call stacks can be printed
  by the interpreter.
* Change the order of error reporting so that it is more clear what
  part of a printout is attributable to the interpreter or shape prop
  adding additional context to an error.
* The first-class functions are only enabled under a guard because
  serialization does not work and more tests need to be added for
  the funnctionality.

gh-metadata: pytorch pytorch 21325 gh/zdevito/48/head

Commit:
Update on "Interpreter support for CallFunction/CallMethod"

Interpreter support for CallFunction/CallMethod

Summary: Adds CALL instruction, and function_table to interpreter state.
Interpreter state now has a list of Frame objects to keep track of
active call frames.

Nodes:
* Re-did how exception wrapping works so that call stacks can be printed
  by the interpreter.
* Change the order of error reporting so that it is more clear what
  part of a printout is attributable to the interpreter or shape prop
  adding additional context to an error.
* The first-class functions are only enabled under a guard because
  serialization does not work and more tests need to be added for
  the funnctionality.

gh-metadata: pytorch pytorch 21325 gh/zdevito/48/head

zdevito committed:
Update on "Interpreter support for CallFunction/CallMethod"

Interpreter support for CallFunction/CallMethod

Summary: Adds CALL instruction, and function_table to interpreter state.
Interpreter state now has a list of Frame objects to keep track of
active call frames.

Nodes:
* Re-did how exception wrapping works so that call stacks can be printed
  by the interpreter.
* Change the order of error reporting so that it is more clear what
  part of a printout is attributable to the interpreter or shape prop
  adding additional context to an error.
* The first-class functions are only enabled under a guard because
  serialization does not work and more tests need to be added for
  the funnctionality.

gh-metadata: pytorch pytorch 21325 gh/zdevito/48/head

zdevito committed:
Update on "Interpreter support for CallFunction/CallMethod"

Interpreter support for CallFunction/CallMethod

Summary: Adds CALL instruction, and function_table to interpreter state.
Interpreter state now has a list of Frame objects to keep track of
active call frames.

Nodes:
* Re-did how exception wrapping works so that call stacks can be printed
  by the interpreter.
* Change the order of error reporting so that it is more clear what
  part of a printout is attributable to the interpreter or shape prop
  adding additional context to an error.
* The first-class functions are only enabled under a guard because
  serialization does not work and more tests need to be added for
  the funnctionality.

gh-metadata: pytorch pytorch 21325 gh/zdevito/48/head
